### PR TITLE
[BOUNTY] MELEE REWORK!

### PR DESCRIPTION
**Pull request by ErisMeleeEnjoyer**
I died for your sins

## About The Pull Request

- [x] Intervention attacks: with a base chance of 30% that scales with the victim's toughness down to 5% on 80 TGH(godlike level), one of these intervention attacks will proc on melee hit:
Bloodstains - clothes of the victim become stained with blood, lowering their style, while the victim loses 5 points of sanity
Out of breath - victim takes 10 oxydamage and 5 agony damage
If the victim has a gun in one of their hands: Discharge - victim's gun shoots in a direction away from the attacker

- [x] Distant attacks: spears have a plus one tile increased range to attack when they are wielded

- [x] Push attacks: when aiming groin on disarm intent, spears and hammers send the victim one tile away from the attacker

- [x] Double tact: With ALL ITEMS that have size of HUGE or BULKY, you have to click twice to attack someone. First click is indicated with an alert and a visual indicator.

- [x] Swing: on harm intent, ALL ITEMS of NORMAL size or above can be used for swinging, attacking in an AoE in front of you if you wield them

- [x] Swing animation, sound and double tact visual indicator.

- [x] Broad strike: all bladed weapons save for spears can can hit three different limbs (one of which you targeted) on disarm intent 

- [X] Forced broad strike: saws have broad strike mode always on

- [x] Halberd: an axe on a long stick, this aberration has a fixed broad strike and swing with double tact, but also increased range

- [x] Melee UI: All new properties will be displayed in the tool info 

- [x] Human shields for active blocking: depending on the grab level, the grab either breaks(passive), turns blocking off after a succesful block attempt(aggressive) or always stays on(neck); passive neckgrab blocking for bullets is unaffected

- [x] Countering grabs with items: When a person grabbing someone takes any melee damage, if they have a grab in one of their hands, with a chance of (damage taken minus their toughness) to lower their grab level, halved if the item is NOT edged and halved further if the item is bulky or higher

Stumble now works only if legs were hit. Attacks for dislocation now switched to Help intent.

## Why It's Good For The Game

I am Eris Melee Enjoyer, I have been paid to do this, I must follow the notes, I must make melee **good**.

## Changelog
:cl: Kegdo The Enjoyer of Melee
add: intervention attacks
add: push attacks
add: distant attacks
add: swing attacks
add: broad strike attacks
add: double tact
add: countering grabs with items
add: melee UI
add: human shields for active blocking
balance: all items of bulky size and bigger have double tact
tweak: stumble now works only if legs were hit. attacks for dislocation now switched to Help intent.
soundadd: swing
imageadd: swing and ready indicator
/:cl:

### CODE REVIEW

**Comment by DiaFRAME444** [comment]
> Distant attacks: spears have a plus one tile increased range to attack when they are wielded

Movement in SS13 is too unwieldy for this to really matter. even if you outspeed someone you'll most likely be dancing in and out of their melee range since slipping from 2 tiles to 1 tile is too easy.

>  Double tact: With ALL ITEMS that have size of HUGE or BULKY, you have to click twice to attack someone. First click is indicated with an alert and a visual indicator.
> Swing: on harm intent, ALL ITEMS of HUGE size can be used for swinging, attacking in an AoE in front of you

SS14 tried this exact melee system with charged AoE attacks and it was crap, for the same reason that movement is unwieldy.

I know this is following a old design document, save that comment. I don't think being a document magically makes the exact same idea work better.

**Comment by Mycah142** [comment]
> > Distant attacks: spears have a plus one tile increased range to attack when they are wielded
> 
> Movement in SS13 is too unwieldy for this to really matter. even if you outspeed someone you'll most likely be dancing in and out of their melee range since slipping from 2 tiles to 1 tile is too easy.
> 
> > Double tact: With ALL ITEMS that have size of HUGE or BULKY, you have to click twice to attack someone. First click is indicated with an alert and a visual indicator.
> > Swing: on harm intent, ALL ITEMS of HUGE size can be used for swinging, attacking in an AoE in front of you
> 
> SS14 tried this exact melee system with charged AoE attacks and it was crap, for the same reason that movement is unwieldy.
> 
> I know this is following a old design document, save that comment. I don't think being a document magically makes the exact same idea work better.

Each of these items were in the design document and aren't up to debate I believe. It's already been approved.

**Comment by maikilangiolo** [comment]
Double tap is shit and will kill every single weapon that needs it. Bulky weapons being worse to counter grabs makes said weapons keep eating shit from grabs, despite being intended to be heavy hitters and dangerous to fight in CQC. 

Just have faster weapons have a lower click cooldown and bulkier being slower, without two click bullshit.

**Comment by SaltyOldBull** [comment]
good luck kegdo

File: code/modules/mob/living/carbon/human/human_attackhand.dm [hunk 1]
```diff
@@ -219,16 +219,21 @@
...
-				//Tell everyone about blocking
-				H.attack_log += text("\[[time_stamp()]\] <font color='orange'>Blocked attack of [src.name] ([src.ckey])</font>")
-				src.attack_log += text("\[[time_stamp()]\] <font color='orange'>Attack has been blocked by [H.name] ([H.ckey])</font>")
-				visible_message(SPAN_WARNING("[src] blocks the blow!"), SPAN_DANGER("You block the blow!"))
-				//They farked up
-				if(real_damage == 0)
-					visible_message(SPAN_DANGER("The attack has been completely negated!"))
+				if(istype(get_active_hand(), /obj/item/grab))//we are blocking with a human shield! We redirect the attack. You know, because grab doesn't exist as an item.
+					var/obj/item/grab/G = get_active_hand()
+					grab_redirect_attack(G)
```
Comment (Pink-Chink) [inline code/modules/mob/living/carbon/human/human_attackhand.dm:224]:
```suggestion
					grab_redirect_attack(M, G)
```
See suggestions in grab_redirect_attack()

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 2]
```diff
@@ -261,6 +228,31 @@ meteor_act
...
+	var/mob/living/carbon/human/attacker = G.assailant
+	var/mob/living/carbon/human/grabbed = G.affecting
+	visible_message(SPAN_DANGER("[src] redirects the blow at [grabbed]!"), SPAN_DANGER("You redirect the blow at [grabbed]!"))
+	//check what we are being hit with, a hand(I is null), or an item?
+	//quickly turn blocking off and on to prevent looping(since we are attacking again)
+	blocking = FALSE
+	if(istype(I, /obj/item))
+		grabbed.attackby(I, attacker)
+	else
+		attacker.attack_hand(grabbed)//and now it's not our problems
```
Comment (Pink-Chink) [inline code/modules/mob/living/carbon/human/human_defense.dm:241]:
```suggestion
		grabbed.attack_hand(attacker)//and now it's not our problems
```
This was the wrong way around.

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 3]
```diff
@@ -261,6 +228,31 @@ meteor_act
...
+	blocking = TRUE
+	//change our block state depending on grab level
+	if(G.state < GRAB_NECK)
+		return //block remains active
+	else if(G.state < GRAB_AGGRESSIVE)
+		stop_blocking()
+		return //block is turned off
+	else
+		stop_blocking()
+		attacker.drop_from_inventory(G)
```
Comment (Pink-Chink) [inline code/modules/mob/living/carbon/human/human_defense.dm:251]:
```suggestion
		drop_from_inventory(G)
```
The mob holding the grabbed mob is the one calling the proc.

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 4]
```diff
@@ -261,6 +228,31 @@ meteor_act
...
+	blocking = FALSE
+	if(istype(I, /obj/item))
+		grabbed.attackby(I, attacker)
+	else
+		attacker.attack_hand(grabbed)//and now it's not our problems
+	blocking = TRUE
+	//change our block state depending on grab level
+	if(G.state < GRAB_NECK)
+		return //block remains active
+	else if(G.state < GRAB_AGGRESSIVE)
```
Comment (Pink-Chink) [inline code/modules/mob/living/carbon/human/human_defense.dm:246]:
```suggestion
	if(G.state >= GRAB_NECK)
		return //block remains active
	else if(G.state >= GRAB_AGGRESSIVE)
```
Checks were backwards. `G.state < GRAB_NECK` was catching the regular grabs and keeping the block up.

**Comment by Alliostra** [comment]
melee attacks should have a minimum chance to lower the level of the grab. Currently, if the grabber has 20 TGH (attainable roundstart as vagabond), you need to deal at least 40 damage per hit for a meager 10% chance to downgrade the grab level. With that kind of chance, the grabber will be long dead before you can break out of his grab this way.

If broad strike really needs to exist, either make it deal more damage in total (so something like 35 or 40% of regular damage per body part), or make it only exist as a forced thing on some weapon types. There's 0 reason to use broad strike if you can also not use broad strike, so it existing as an optional thing is just a noob trap.

Also, obligatory "Double tact should not exist"

**Comment by ErisMeleeEnjoyer** [comment]
Should be ready to merge. All bugs from test server have been adressed.

**Comment by ValoTheValo** [comment]
remove double tap. its bad and adds nothing

**Review by Firefox13 — COMMENTED** [review]
Double-tact isn't still really as Reere doc says, citing `"You can just click on your target regardless of distance, it will still make you raise your weapon and prepare for attack, allowing you to do that before going to melee range."`
This isn't a thing right now and it's fucking shitty. Perhaps for first click (weapon 'raising') you can click literally anywhere?

**Comment by DiaFRAME444** [comment]
> Double-tact isn't still really as Reere doc says, citing `"You can just click on your target regardless of distance, it will still make you raise your weapon and prepare for attack, allowing you to do that before going to melee range."` This isn't a thing right now and it's fucking shitty. Perhaps for first click (weapon 'raising') you can click literally anywhere?

Wait, literally what is the point of that implementation?
People can already see that you're suddenly wielding a huge melee weapon because all melees have notable sprites.

**Comment by BlackjackIsTired** [comment]
If Double-tact is staying, I would suggest finding another indicator for showing someone's weapon is ready and prepared to strike. As it is now the large red floating triangle feels extremely arcadey and bombs the Eris art style and atmosphere.

**Comment by ErisMeleeEnjoyer** [comment]
> Double-tact isn't still really as Reere doc says, citing `"You can just click on your target regardless of distance, it will still make you raise your weapon and prepare for attack, allowing you to do that before going to melee range."` This isn't a thing right now and it's fucking shitty. Perhaps for first click (weapon 'raising') you can click literally anywhere?

Done. That was hell.

**Comment by Pink-Chink** [comment]
Good to merge. Bounty completion is up to Firefox.

**Pull request merged**